topic-operator/src/main/java/io/strimzi/operator/topic/v2/BatchingLoop.java

```diff
@@ -0,0 +1,302 @@
+ * the reconciliation of those events using a {@link BatchingTopicController}.
+ * Any given KafkaTopic is only being reconciled by a single thread at any one time.
+ */
+class BatchingLoop {
+
+    static final ReconciliationLogger LOGGER = ReconciliationLogger.create(BatchingLoop.class);
+
+    private final BatchingTopicController controller;
+    private final BlockingDeque<TopicEvent> queue;
+
+    /**
+     * The set of topics currently being reconciled by a controller.
+     * Guarded by the monitor of the BatchingLoop.
+     * This functions as mechanism for preventing concurrent reconciliation of the same topic.
+     */
+    private final Set<Ref> inFlight = new HashSet<>(); // guarded by this
+    private final LoopRunnable[] threads;
+    private final int maxBatchSize;
+    private final long maxBatchLingerMs;
+    private final ItemStore<KafkaTopic> itemStore;
+
+    public BatchingLoop(
+            int maxQueueSize,
+            BatchingTopicController controller,
+            int maxThreads,
+            int maxBatchSize,
+            long maxBatchLingerMs,
+            ItemStore<KafkaTopic> itemStore) {
+        this.queue = new LinkedBlockingDeque<>(maxQueueSize);
+        this.controller = controller;
+        this.threads = new LoopRunnable[maxThreads];
+        for (int i = 0; i < maxThreads; i++) {
+            threads[i] = new LoopRunnable("LoopRunnable-" + i);
+        }
+        this.maxBatchSize = maxBatchSize;
+        this.maxBatchLingerMs = maxBatchLingerMs;
+        this.itemStore = itemStore;
+    }
+
+    /**
+     * Starts the threads
+     */
+    public void start() {
+        for (var thread : threads) {
+            thread.start();
+        }
+    }
+
+    /**
+     * Stops the threads
+     * @throws InterruptedException If interrupted while waiting for the threads to stop.
+     */
+    public void stop() throws InterruptedException {
+        for (var thread : threads) {
+            thread.requestStop();
+        }
+        for (var thread : threads) {
+            thread.join();
+        }
+    }
+
+    /**
+     * Add an event to be reconciled to the tail of the {@link #queue}.
+     * @param event The event
+     */
+    public void offer(TopicEvent event) {
+        if (queue.offerFirst(event)) {
+            LOGGER.debugOp("Item {} push onto the deque tail", event);
+        } else {
+            LOGGER.warnOp("Item {} could not be pushed onto the deque tail", event);
+        }
+    }
+
+    /**
+     * The loop is alive if none of the threads have been blocked for more than 2 minutes.
+     * "Blocked" means they're not returned to their outermost loop.
+     * @return True if the loop is alive..
+     */
+    boolean isAlive() {
+        for (var t : threads) {
+            if (!t.isAlive()) {
+                LOGGER.infoOp("isAlive returning false because {} is not alive", t);
+                return false;
+            } else if (t.msSinceLastLoop() > 120_000L) {
```

Comment:
Maybe they are too low level parameters, but from where 2 mins come? Why it's not 1, 3, 5? Just trying to understand the rational behind that number.

Reply:
It didn't come from any principled argument. 1 seemed too small, and 3 seemed too long. What are the consequences of getting it wrong? Only that the TO will be restarted when it _might have_ recovered, or won't be restarted as quickly as it would have when it might not have recovered. I.e. even with the above logic added there would be no correct value.